dropshot/src/server.rs

```diff
@@ -578,15 +645,84 @@ impl<C: ServerContext> HttpServer<C> {
     }
 
     /**
-     * Signals the currently running server to stop and waits for it to exit.
+     * Return the result of registering the server's DTrace USDT probes.
+     *
+     * See [`ProbeRegistration`] for details.
      */
+    pub fn probe_registration(&self) -> &ProbeRegistration {
+        &self.probe_registration
+    }
+}
+
+impl<C: ServerContext, CL: Closer> HttpServer<C, CL, Joinable> {
+    /// Separates the [HttpServer] from the [HttpServerJoining] object which is
+    /// capable of joining it.
+    pub fn get_joiner(
+        self,
+    ) -> (HttpServer<C, CL, Unjoinable>, HttpServerJoiner) {
+        let HttpServer {
+            probe_registration,
+            app_state,
+            local_addr,
+            closer,
+            joiner,
+            _phantom: PhantomData,
+        } = self;
+
+        (
+            HttpServer {
+                probe_registration,
+                app_state,
+                local_addr,
+                closer,
+                joiner: None,
+                _phantom: PhantomData,
+            },
+            joiner.unwrap(),
+        )
+    }
+}
+
+impl<C: ServerContext, J: Joiner> HttpServer<C, Closeable, J> {
+    /// Separates the [HttpServer] from the [HttpServerCloser] object which is
+    /// capable of closing it.
+    pub fn get_closer(self) -> (HttpServer<C, Uncloseable>, HttpServerCloser) {
+        let HttpServer {
+            probe_registration,
+            app_state,
+            local_addr,
+            closer,
+            joiner,
+            _phantom: PhantomData,
+        } = self;
+
+        (
+            HttpServer {
+                probe_registration,
+                app_state,
+                local_addr,
+                closer: None,
+                joiner,
+                _phantom: PhantomData,
+            },
+            closer.unwrap(),
+        )
+    }
+}
+
+impl<C: ServerContext> HttpServer<C, Closeable, Joinable> {
+    /// Signals the currently running server to stop and waits for it to exit.
     pub async fn close(mut self) -> Result<(), String> {
```

Comment:
This function is really doing "close and join", but for backwards compatibility, it's just called "close".